packages/caver-wallet/src/keyring/keyring.js

```diff
@@ -0,0 +1,845 @@
+        if (this._key[KEY_ROLE.ROLE_TRANSACTION_KEY].length > 1) throw new Error(notAvailableError)
+
+        for (let i = KEY_ROLE.ROLE_ACCOUNT_UPDATE_KEY; i < KEY_ROLE.ROLE_LAST; i++) {
+            if (this._key[i].length > 0) {
+                throw new Error(notAvailableError)
+            }
+        }
+
+        options = options || {}
+
+        const crypto = encryptKey(this._key[0][0], password, options)
+
+        return formatEncrypted(3, this._address, crypto, options)
+    }
+
+    _validateOptionsForUpdate(options = []) {
+        // { threshold: 1, weight: [1, 1] } => [{ threshold: 1, weight: [1, 1] }]
+        if (!_.isArray(options)) options = [options]
+
+        for (let i = 0; i < this._key.length; i++) {
+            // Validation for options obejct will be operated in AccountKeyWeightedMultiSig class.
+            if (options[i] && Object.keys(options[i]).length > 0) continue
+
+            let optionToAdd
+            if (this._key[i].length > 1) {
+                // default option when option is not set
+                optionToAdd = { threshold: 1, weight: Array(this._key[i].length).fill(1) }
+            } else {
+                // AccountKeyPublic does not need option
+                optionToAdd = {}
+            }
+
+            if (options[i]) {
+                options[i] = optionToAdd
+            } else {
+                options.push(optionToAdd)
+            }
+        }
+        return options
+    }
+}
+
+Keyring.PrivateKey = PrivateKey
+Keyring.role = KEY_ROLE
+
+module.exports = Keyring
+
+// PrivateKey instance: PrivateKey{}
+// single private key string: `0x{private key}`
+// multiple private key strings: [`0x{private key}`, `0x{private key}`]
+// multiple PrivateKey instances: [PrivateKey{}, PrivateKey{}]
+// role-based private key strings: [[`0x{private key}`], [`0x{private key}`, `0x{private key}`], [`0x{private key}`]]
+// role-based PrivateKey instances: [[PrivateKey{}], [PrivateKey{}, PrivateKey{}], [PrivateKey{}]]
+function formattingForKeyInKeyring(keyInput) {
+    if (keyInput === null) {
+        return keyInput
+    }
+
+    if (keyInput instanceof PrivateKey || _.isString(keyInput)) {
+        keyInput = [[keyInput], [], []]
+    } else if (isMultipleKeysFormat(keyInput)) {
+        // [`0x{private key}`, `0x{private key}`, `0x{private key}`]
+        keyInput = [keyInput, [], []]
+    } else if (!isRoleBasedKeysFormat(keyInput)) {
+        throw new Error(`Invalid format for key variable in keyring`)
+    }
+
+    const keys = generateKeysFormat()
+    for (let i = 0; i < KEY_ROLE.ROLE_LAST; i++) {
+        fillRoleKey(keys, i, keyInput[i])
+    }
+
+    return keys
+}
+
+function generateKeysFormat() {
+    return Array(KEY_ROLE.ROLE_LAST)
+        .fill(null)
+        .map(() => [])
+}
+
+function isMultipleKeysFormat(keys) {
+    if (!_.isArray(keys)) return false
+    return keys.every(key => {
+        return _.isString(key)
```

Comment:
Don't we need to check the validity of each private key here?

Reply:
This function is used to check if the keys entered as parameters are in the format for multiple keys. (['0x..', '0x...', '0x...'] format)
There is no need to validate the private key here, and validation is performed in the set function of the private key.

And also i added limitation of the number of private key in 'fillRoleKey' :)